### ISSUE TITLE
intel compiler: nonstandard use of "auto" to both deduce the type from an initializer and to announce a trailing return type

### ISSUE THREAD

**Issue by heavytail** [pybind/pybind11#932]
With intel compiler 17.0.1, compiling the following sample file with -std=c++14
```
#include <pybind11/pybind11.h>
#include <pybind11/numpy.h>

int main() {
    return 0;
}
```
will produce warning:
```
/home/foo/.local/include/python2.7/pybind11/cast.h(1815): warning #3373: nonstandard use of "auto" to both deduce the type from an initializer and to announce a trailing return type
          static constexpr auto args_pos = constexpr_first<argument_is_args, Args...>() - (int) sizeof...(Args),
                                                                                                                 ^
```
It is due to intel compiler considers `auto` for initializing two variables `args_pos` and `kwargs_pos` in the same line as nonstandard:
```
static constexpr auto args_pos = constexpr_first<argument_is_args, Args...>() - (int) sizeof...(Args),
                    kwargs_pos = constexpr_first<argument_is_kwargs, Args...>() - (int) sizeof...(Args);
```

This warning can be cleared by define `args_pos` and `kwargs_pos` in separate lines:
```
static constexpr auto args_pos = constexpr_first<argument_is_args, Args...>() - (int) sizeof...(Args);
static constexpr auto kwargs_pos = constexpr_first<argument_is_kwargs, Args...>() - (int) sizeof...(Args);
```

**Comment by jagerman** [comment]
This seems more like an intel compiler bug; declaring multiple variables with `auto` is definitely allowed by the standard as long as the variables are the same type (here both expressions are unequivocally `int`s).  I'm more tempted to just add a pragma to disable warning 3373 under icc to avoid the false positive warning.

**Comment by dean0x7d** [comment]
I just compiled with icc 17.0.4 on Linux and it doesn't report this warning, so I assume it has been fixed in the compiler at some point since 17.0.1.

(icc 17.0.4 can't actually compile `master` at the moment. I'll submit a fix shortly.)

**Comment by dean0x7d** [comment]
PR #937 fixes an Intel C++ compiler error on `master` and resolves a few additional warnings. In light of that and the icc bugfix update which resolves this specific warning, I think that the issues related to Intel C++ on Linux are resolved for now.
